Replace hand-written informers with generated ones

ncdc opened the pull request:
Replace existing uses of hand-written informers with generated ones.
Follow-up commits will switch the use of one-off informers to shared
informers.

This is a precursor to #40097. That PR will switch one-off informers to shared informers for the majority of the code base (but not quite all of it...).

NOTE: this does create a second set of shared informers in the kube-controller-manager. This will be resolved back down to a single factory once #40097 is reviewed and merged.

There are a couple of places where I expanded the # of caches we wait for in the calls to `WaitForCacheSync` - please pay attention to those. I also added in a commented-out wait in the attach/detach controller. If @kubernetes/sig-storage-pr-reviews is ok with enabling the waiting, I'll do it (I'll just need to tweak an integration test slightly).

@deads2k @sttts @smarterclayton @liggitt @soltysh @timothysc @lavalamp @wojtek-t @gmarek @sjenning @derekwaynecarr @kubernetes/sig-scalability-pr-reviews

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40385)
<!-- Reviewable:end -->

smarterclayton commented:
I didn't spot anything else that surprised me except the two comments.  Is the quota failure related?

smarterclayton commented:
/approve

@wojtek-t iirc this may temporarily regress memory usage and maybe CPU on kubemark-500 - the end result after the other commit should be back to below.

smarterclayton commented:
That was iiuc

jsafrane commented:
PV controller needs to get periodic sync quite often (currently 15 seconds) to keep reasonable time for PV-PVC binding. Is this possible with a shared controller? #26406 is still not fixed.

ncdc commented:
It's not currently possible but it is something we'd like to fix asap. I
can back out the PV controller from this change for now.

cc @deads2k

On Wed, Jan 25, 2017 at 4:05 AM Jan Šafránek <notifications@github.com>
wrote:

> PV controller needs to get periodic sync quite often (currently 15
> seconds) to keep reasonable time for PV-PVC binding. Is this possible with
> a shared controller? #26406
> <https://github.com/kubernetes/kubernetes/issues/26406> is still not
> fixed.
>
>
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/40385#issuecomment-275054436>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYjSsR39Q9q-FOSnldCxxvezy7_x1ks5rVxBMgaJpZM4LszVS>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

ncdc commented:
@jsafrane this PR doesn't touch the PV controller; it only touches attach/detach.

jsafrane commented:
I think attach/detach should be fine with shared informers, cc @saad-ali and @jingxu97 to confirm that.

ncdc commented:
It's already using them. This PR just switches to use the generated ones.

On Wed, Jan 25, 2017 at 7:08 AM Jan Šafránek <notifications@github.com>
wrote:

> I think attach/detach should be fine with shared informers, cc @saad-ali
> <https://github.com/saad-ali> and @jingxu97 <https://github.com/jingxu97>
> to confirm that.
>
>
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/40385#issuecomment-275092157>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYqYVR_IZQF4-8OXiq1khBpZ1Gtfoks5rVztHgaJpZM4LszVS>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

ncdc commented:
@smarterclayton pushed an update that removes my test hacking. I'm not sure about the quota failures. Let's see what happens in this next round of testing.

ncdc commented:
Ok the quota tests are failing repeatedly. Either something is up in this PR, or test-infra. I'll start examining the PR

ncdc commented:
@smarterclayton it looks like the quota issue was because 1 portion of the resource quota controller was using the legacy shared informers, and another part was using the new ones. I've reverted that change so the quota controller is only using legacy. I'll switch it to new in the next PR.

ncdc commented:
Follow-up issue to ensure controllers wait for caches to sync: https://github.com/kubernetes/kubernetes/issues/40444

smarterclayton commented:
Lgtm but needs rebase

ncdc commented:
Will do, but I want to make sure we can figure out per-listener resync
periods against a shared informer first, otherwise we can't really do phase
2 (convert the non-shared informers to shared).

On Wed, Jan 25, 2017 at 8:51 PM Clayton Coleman <notifications@github.com>
wrote:

> Lgtm but needs rebase
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/40385#issuecomment-275288624>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYrshsR8kXpLgfZ-6SX1wYQClvA6_ks5rV_wZgaJpZM4LszVS>
> .
>

deads2k commented on pkg/controller/deployment/deployment_controller.go:
```diff
@@ -495,17 +498,20 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 		return nil
 	}
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Unable to retrieve deployment %v from store: %v", key, err))
```
I see its pre-existing, but we need to sweep and eliminate double reporting

deads2k commented on pkg/controller/replication/replication_controller.go:
```diff
@@ -678,14 +664,14 @@ func (rm *ReplicationManager) syncReplicationController(key string) error {
 	}
 	trace.Step("manageReplicas done")
 
 	newStatus := calculateStatus(rc, filteredPods, manageReplicasErr)
```
point me to the copy you make before calling here?

ncdc replied:
arg, it didn't make it over from the other branch. i'll bring it back :)

deads2k replied:
> arg, it didn't make it over from the other branch. i'll bring it back :)

did this get resolved?  If so, you're just a rebase away from merging, right?

ncdc replied:
@deads2k I don't really want to merge until after we're close to agreement on how the per-listener resync periods will work for shared informers. If we can't get that, this will have to wait. I'm still wrapping that up.

deads2k commented:
Once question I'd block on.  lgtm otherwise.

@eparis we'll need approval again.  I should probably be allowed to approve this one.

eparis commented:
@deads2k I'd probably harass @smarterclayton to /approve

ncdc commented:
No need to approve until I've rebased and fixed the missing copy bit

deads2k commented:
> @deads2k I'd probably harass @smarterclayton to /approve

For these packages, can you add me.  I've got a lot of commits and reviews in each of them. pkg/controllers, cmd, plugin/pkg/admission.

mikedanese commented:
@apelisse why did merge bot remove @smarterclayton 's approve?

mikedanese commented:
/approve

ncdc commented:
Perhaps because I pushed an update after the initial approval?

smarterclayton commented:
/approve

On Thu, Jan 26, 2017 at 2:50 PM, Andy Goldstein <notifications@github.com>
wrote:

> Perhaps because I pushed an update after the initial approval?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/40385#issuecomment-275493722>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pwxoYs-MnriegM189-MiIvXGEwT7ks5rWPj5gaJpZM4LszVS>
> .
>

deads2k commented:
> @deads2k I don't really want to merge until after we're close to agreement on how the per-listener resync periods will work for shared informers. If we can't get that, this will have to wait. I'm still wrapping that up.

How does this pull make that problem worse?

ncdc commented:
Because it creates a temporary second set of shared informers. I don't want to do that until I know that we can resolve the per-listener resync in time for 1.6, as all the code that we want to migrate from non-shared to shared informers will need custom resyncs.

ncdc commented:
@deads2k rebased and hopefully added back in the deep copying we need for rc/rs. ptal

deads2k commented:
@k8s-bot kops aws e2e test this

deads2k commented:
/lgtm

deads2k commented:
/lgtm

ncdc commented:
I need to back out some changes to the resource quota controller before this can merge. I could have sworn I did that already, although maybe it was in a different branch...

Commit:
Replace hand-written informers with generated ones

 Replace existing uses of hand-written informers with generated ones.
 Follow-up commits will switch the use of one-off informers to shared
 informers.

ncdc commented:
Ok I think I've restored the branch to the state it needs to be in (not modifying quota for now).

deads2k commented:
/lgtm

derekwaynecarr commented:
/approve

ncdc commented:
@k8s-bot cvm gke e2e test this

deads2k commented:
/approve

smarterclayton commented:
/approve

smarterclayton commented:
/approve

On Mon, Feb 6, 2017 at 3:01 PM, Kubernetes Submit Queue <
notifications@github.com> wrote:

> [APPROVALNOTIFIER] This PR is *NOT APPROVED*
>
> The following people have approved this PR: *deads2k, derekwaynecarr,
> ncdc*
>
> Needs approval from an approver in each of these OWNERS Files:
>
>    - *cmd/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/cmd/OWNERS>*
>    - cmd/kube-controller-manager/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/OWNERS>
>    [deads2k]
>    - pkg/controller/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/OWNERS>
>    [derekwaynecarr]
>    - pkg/controller/daemon/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/OWNERS>
>    [derekwaynecarr]
>    - pkg/controller/deployment/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/OWNERS>
>    [derekwaynecarr]
>    - pkg/controller/job/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/OWNERS>
>    [derekwaynecarr]
>    - pkg/controller/node/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/node/OWNERS>
>    [derekwaynecarr]
>    - pkg/controller/podgc/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podgc/OWNERS>
>    [derekwaynecarr]
>    - pkg/controller/replicaset/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/OWNERS>
>    [derekwaynecarr]
>    - pkg/controller/replication/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replication/OWNERS>
>    [derekwaynecarr]
>    - pkg/controller/volume/attachdetach/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/OWNERS>
>    [derekwaynecarr]
>    - test/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/test/OWNERS>
>    [deads2k]
>
> We suggest the following people:
> cc @mikedanese <https://github.com/mikedanese>
> You can indicate your approval by writing /approve in a comment
> You can cancel your approval by writing /approve cancel in a comment
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/40385#issuecomment-277795761>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p908ZLb4RofQ5yCp0QQ8O2yvlo5cks5rZ3wegaJpZM4LszVS>
> .
>
